core/src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala

```diff
@@ -407,7 +438,7 @@ case class MergeIntoCommand(
 
     // UDF to increment metrics
     val incrSourceRowCountExpr = makeMetricUpdateUDF("numSourceRows")
-    val sourceDF = Dataset.ofRows(spark, source)
+    val sourceDF = sourcePlanDF
```

Comment:
this is seems fairly risky use for `var` which is initialized to null. rather than risk a NPE, it might be better to wrap this into a method like `getSourceDF` (which also indicates that this is a method, not a local variable) where you can throw a better error message if somehow the var is not initialized correctly to null.